docs/prober-and-monitoring.md

```diff
@@ -0,0 +1,102 @@
+# Prober and monitoring
+
+**PMAP is not a official Google product.**
+
+Monitoring consists of two parts: prober and alerting.
+
+## Prober
+
+[Prober](../prober/) is a tool that constantly probes pmap services to check if
+the services are up. It is built as a go binary and deployed to [cloud run
+job](https://cloud.google.com/run/docs/overview/what-is-cloud-run#jobs), and is
+triggered by [cloud scheduler](https://cloud.google.com/scheduler) to constantly
+probing pmap services to check if the services are up.
+
+In each prober execution, the prober will cover two CUJ:
+
+- Import resource mapping to BigQuery from GCS
+- Import policies to BigQuery from GCS
+
+For each CUJ, prober will upload a yaml file to
+[gcs](https://cloud.google.com/storage), and query the bigquery table to see if
+a corrosponding entry exists in bigquery table. The job is considered as success
+on if the following requirements for both mapping and policy service are met:
+
+1. The object is uploaded.
+2. There is an entry in bigquery table that matches the traceID.
+
+### Limitation
+
+Ideally the prober should cover the full user journey which includes checking in
+resource mappings and policies in a GitHub repository. However, there is no
+effective way to automate PR authoring and bypass branch protection. As a result,
+we seek the next closest journey - upload probing files to GCS bucket directly.
+
+## Monitoring and Alert Policies
+
+We also use [cloud run
+metircs](https://cloud.google.com/monitoring/api/metrics_gcp#gcp-run) to monitor
+and send alert base prober job execution result. The metric used is:
+
+- job/completed_execution_count
+
+We also use [pubsub
+metrics](https://cloud.google.com/monitoring/api/metrics_gcp#gcp-pubsub) to
+monitor pubsub subsrciption and sent out alerts. Pubsub messages plays a vital role in pmap services, as they are used to trigger mapping and policy
+services, and write processing result into bigquery. The metrics used are:
+
+- subscription/oldest_unacked_message_age
+- subscription/dead_letter_message_counts
+
+## Installation
+
+### Default setup
+
+If you used the e2e module to set up pmap service, then you don't need to set up
+prober separately. Otherwise, you can use the following terraform code to set up
+your prober. All these variables are required.
+
+```terraform
+module "prober_and_monitoring" {
+  source = "../modules/monitoring"
+
+  project_id = var.project_id
+
+  prober_bucket_id           = "GCS bucket id for where objects are uploaded to."
```

Comment:
prober_bucket_id           = "<YOUR_GCS_bucket_id_for_where_objects_are_uploaded to>" Just wanna make it clear that what variables that users need to provide.

Reply:
done